File: packages/core-python/jsonmd/mcp.py
"""MCP schema compression utilities."""

from __future__ import annotations

from typing import Any


def is_mcp_tool_list(data: Any) -> bool:
    """Detect if a value is an MCP tool list response."""
    if not isinstance(data, dict):
        return False
    tools = data.get("tools")
    if not isinstance(tools, list) or len(tools) == 0:
        return False
    return all(
        isinstance(t, dict)
        and isinstance(t.get("name"), str)
        and (isinstance(t.get("inputSchema"), dict) or isinstance(t.get("parameters"), dict))
        for t in tools
    )


def _get_schema(tool: dict) -> dict:
    return tool.get("inputSchema") or tool.get("parameters") or {}


def _type_str(prop: dict) -> str:
    t = prop.get("type", "unknown")
    if isinstance(t, list):
        return " | ".join(t)
    if t == "array" and isinstance(prop.get("items"), dict):
        item_type = prop["items"].get("type", "unknown")
        return f"{item_type}[]"
    return t


def _mcp_to_table(tools: list[dict]) -> str:
    lines = []
    lines.append("# Tools")
    lines.append("")
    lines.append("| Tool | Description |")
    lines.append("| --- | --- |")
    for tool in tools:
        lines.append(f"| {tool['name']} | {tool.get('description', '')} |")

    for tool in tools:
        schema = _get_schema(tool)
        props = schema.get("properties", {})
        if not props:
            continue

        required = set(schema.get("required", []))
        lines.append("")
        lines.append(f"## {tool['name']}")
        lines.append("")
        if tool.get("description"):
            lines.append(tool["description"])
            lines.append("")
        lines.append("| Param | Type | Req | Description |")
        lines.append("| --- | --- | --- | --- |")
        for name, prop in props.items():
            req = "✓" if name in required else ""
            lines.append(f"| {name} | {_type_str(prop)} | {req} | {prop.get('description', '')} |")

    return "\n".join(lines)


def _mcp_to_signatures(tools: list[dict]) -> str:
    lines = []
    lines.append("# Tools")
    lines.append("")

    for tool in tools:
        schema = _get_schema(tool)
        props = schema.get("properties", {})
        required = set(schema.get("required", []))

        params = []
        for name, prop in props.items():
            optional = "" if name in required else "?"
            params.append(f"{name}{optional}: {_type_str(prop)}")

        sig = f"{tool['name']}({', '.join(params)})"
        desc = f" — {tool['description']}" if tool.get("description") else ""
        lines.append(f"- `{sig}`{desc}")

    return "\n".join(lines)


def mcp_to_md(data: dict, mode: str = "table") -> str:
    """Convert MCP tool list to Markdown."""
    tools = data.get("tools", [])
    if mode == "signatures":
        return _mcp_to_signatures(tools)
    return _mcp_to_table(tools)

File: packages/core-js/src/mcp.ts
import type { McpTool, McpSchema, McpProperty } from './types.js';

/**
 * Detect if a value is an MCP tool list response.
 */
export function isMcpToolList(data: unknown): data is { tools: McpTool[] } {
  if (typeof data !== 'object' || data === null || Array.isArray(data)) return false;
  const obj = data as Record<string, unknown>;
  if (!Array.isArray(obj.tools)) return false;
  const tools = obj.tools;
  if (tools.length === 0) return false;
  return tools.every(
    (t: unknown) =>
      typeof t === 'object' &&
      t !== null &&
      typeof (t as Record<string, unknown>).name === 'string' &&
      (typeof (t as Record<string, unknown>).inputSchema === 'object' ||
        typeof (t as Record<string, unknown>).parameters === 'object'),
  );
}

function getSchema(tool: McpTool): McpSchema {
  return tool.inputSchema || tool.parameters || {};
}

function typeStr(prop: McpProperty): string {
  if (Array.isArray(prop.type)) {
    return prop.type.join(' | ');
  }
  if (prop.type === 'array' && prop.items?.type) {
    return `${prop.items.type}[]`;
  }
  return prop.type || 'unknown';
}

/**
 * Convert MCP tool list to Markdown table format.
 */
function mcpToTable(tools: McpTool[]): string {
  const lines: string[] = [];

  // Summary table
  lines.push('# Tools');
  lines.push('');
  lines.push('| Tool | Description |');
  lines.push('| --- | --- |');
  for (const tool of tools) {
    lines.push(`| ${tool.name} | ${tool.description || ''} |`);
  }

  // Per-tool parameter tables
  for (const tool of tools) {
    const schema = getSchema(tool);
    const props = schema.properties;
    if (!props || Object.keys(props).length === 0) continue;

    const required = new Set(schema.required || []);

    lines.push('');
    lines.push(`## ${tool.name}`);
    lines.push('');
    if (tool.description) {
      lines.push(tool.description);
      lines.push('');
    }
    lines.push('| Param | Type | Req | Description |');
    lines.push('| --- | --- | --- | --- |');
    for (const [name, prop] of Object.entries(props)) {
      const req = required.has(name) ? '✓' : '';
      lines.push(`| ${name} | ${typeStr(prop)} | ${req} | ${prop.description || ''} |`);
    }
  }

  return lines.join('\n');
}

/**
 * Convert MCP tool list to compact function signatures.
 */
function mcpToSignatures(tools: McpTool[]): string {
  const lines: string[] = [];
  lines.push('# Tools');
  lines.push('');

  for (const tool of tools) {
    const schema = getSchema(tool);
    const props = schema.properties || {};
    const required = new Set(schema.required || []);

    const params: string[] = [];
    for (const [name, prop] of Object.entries(props)) {
      const optional = required.has(name) ? '' : '?';
      params.push(`${name}${optional}: ${typeStr(prop)}`);
    }

    const sig = `${tool.name}(${params.join(', ')})`;
    const desc = tool.description ? ` — ${tool.description}` : '';
    lines.push(`- \`${sig}\`${desc}`);
  }

  return lines.join('\n');
}

/**
 * Convert MCP tool list to Markdown.
 */
export function mcpToMd(data: { tools: McpTool[] }, mode: 'table' | 'signatures' = 'table'): string {
  return mode === 'signatures' ? mcpToSignatures(data.tools) : mcpToTable(data.tools);
}
